File: celos-server/src/main/java/com/collective/celos/Workflow.java
/*
 * Copyright 2015 Collective, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 */
package com.collective.celos;

import com.collective.celos.trigger.Trigger;

import java.time.ZonedDateTime;

/**
 * A periodical task.
 */
public class Workflow {

    public static final ZonedDateTime DEFAULT_START_TIME = ZonedDateTime.parse("1970-01-01T00:00Z");
    public static final int DEFAULT_WAIT_TIMEOUT_SECONDS = Integer.MAX_VALUE;
    
    private final WorkflowID id;
    private final Schedule schedule;
    private final SchedulingStrategy schedulingStrategy;
    private final Trigger trigger;
    private final ExternalService externalService;
    private final int maxRetryCount;
    private final ZonedDateTime startTime;
    private final int waitTimeoutSeconds;
    private final WorkflowInfo workflowInfo;
    
    public Workflow(WorkflowID id,
                    Schedule schedule,
                    SchedulingStrategy strategy,
                    Trigger trigger,
                    ExternalService service,
                    int maxRetryCount,
                    ZonedDateTime startTime,
                    int waitTimeoutSeconds,
                    WorkflowInfo workflowInfo) {
        this.id = Util.requireNonNull(id);
        this.schedule = Util.requireNonNull(schedule);
        this.schedulingStrategy = Util.requireNonNull(strategy);
        this.trigger = Util.requireNonNull(trigger);
        this.externalService = Util.requireNonNull(service);
        this.maxRetryCount = maxRetryCount;
        this.startTime = Util.requireNonNull(startTime);
        this.workflowInfo = Util.requireNonNull(workflowInfo);
        this.waitTimeoutSeconds = waitTimeoutSeconds;
    }

    public WorkflowID getID() {
        return id;
    }

    public Schedule getSchedule() {
        return schedule;
    }

    public SchedulingStrategy getSchedulingStrategy() {
        return schedulingStrategy;
    }

    public Trigger getTrigger() {
        return trigger;
    }
    
    public ExternalService getExternalService() {
        return externalService;
    }

    public int getMaxRetryCount() {
        return maxRetryCount;
    }

    public ZonedDateTime getStartTime() {
        return startTime;
    }

    public WorkflowInfo getWorkflowInfo() {
        return workflowInfo;
    }

    public int getWaitTimeoutSeconds() {
        return waitTimeoutSeconds;
    }
}

File: celos-server/src/test/resources/com/collective/celos/dependent-slots-test/workflow.js
/*
 * Copyright 2015 Collective, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 */
importPackage(Packages.com.collective.celos);

celos.defineWorkflow({
    "id": "workflow-1",
    "schedule": celos.hourlySchedule(),
    "schedulingStrategy": celos.serialSchedulingStrategy(),
    "trigger": celos.alwaysTrigger(),
    "externalService": new WorkflowConfigurationParserTest$TestExternalService()
});

celos.defineWorkflow({
    "id": "workflow-2",
    "schedule": celos.hourlySchedule(),
    "schedulingStrategy": celos.serialSchedulingStrategy(),
    "trigger": celos.successTrigger("workflow-1"),
    "externalService": new WorkflowConfigurationParserTest$TestExternalService()
});


celos.defineWorkflow({
    "id": "workflow-3",
    "schedule": celos.hourlySchedule(),
    "schedulingStrategy": celos.serialSchedulingStrategy(),
    "trigger": celos.andTrigger(celos.successTrigger("workflow-1"), celos.successTrigger("workflow-2")),
    "externalService": new WorkflowConfigurationParserTest$TestExternalService()
});

celos.defineWorkflow({
    "id": "workflow-4",
    "schedule": celos.hourlySchedule(),
    "schedulingStrategy": celos.serialSchedulingStrategy(),
    "trigger": celos.alwaysTrigger(),
    "externalService": new WorkflowConfigurationParserTest$TestExternalService()
});


var hour_sec = 60*60;

celos.defineWorkflow({
    "id": "workflow-5",
    "schedule": celos.hourlySchedule(),
    "schedulingStrategy": celos.serialSchedulingStrategy(),
    "trigger": celos.orTrigger(celos.notTrigger(celos.successTrigger("workflow-1")), celos.offsetTrigger(hour_sec, celos.successTrigger("workflow-4"))),
    "externalService": new WorkflowConfigurationParserTest$TestExternalService()
});

celos.defineWorkflow({
    "id": "workflow-6-daily",
    "schedule": celos.cronSchedule("0 0 0 * * ?"),
    "schedulingStrategy": celos.serialSchedulingStrategy(),
    "trigger": celos.notTrigger(celos.alwaysTrigger()),
    "externalService": new WorkflowConfigurationParserTest$TestExternalService()
});

celos.defineWorkflow({
    "id": "workflow-7-depend-daily",
    "schedule": celos.hourlySchedule(),
    "schedulingStrategy": celos.serialSchedulingStrategy(),
    "trigger": celos.successTrigger("workflow-6-daily"),
    "externalService": new WorkflowConfigurationParserTest$TestExternalService()
});

